Mobile ucr filters

nickpell opened the pull request:
supersedes https://github.com/dimagi/commcare-hq/pull/7700

@orangejenny @czue

Commit:
pull mobile UCRs from apps instead of entire domain

Commit:
squash me

Commit:
add framework for filtering

Commit:
support choice lists

Commit:
use different schema for different filter types

Commit:
support static date filters

Commit:
support user_id in auto filter

Commit:
support username in auto filter

Commit:
support location id in auto filter

Commit:
support custom user data property in auto filter

Commit:
fix typo

Commit:
WIP

Commit:
support AutoFilter in UI

Commit:
use select for auto filter options

Commit:
add UI for static datespan filter

Commit:
wrap custom data filters

Commit:
improved error message

Commit:
support custom data auto filter in UI

Commit:
implementing static choice filter (WIP)

Commit:
get dynamic choice working in UI

Commit:
save dynamic choice filters in correct format

Commit:
refactor

Commit:
implement select filter

Commit:
make save button responsive to filter UI

Commit:
correct data model for static daterange

Commit:
implement UI for proper daterange

Commit:
python lint

Commit:
js lint

Commit:
refactor

Commit:
more cleanup

Commit:
remove print

Commit:
don't hardcode URL

Commit:
don't use unnecessary observable

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,96 @@ var ReportModule = (function () {
+
+                filter.dynamicFilterName = ko.computed(function () {
+                    return reportId() + '/' + filter.slug;
+                });
```
Don't make functions within a loop.

nickpell commented:
:white_check_mark: Would do a late-night deploy if I can get this merged

czue commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -1,24 +1,76 @@
 {% block js-inline %}{{ block.super }}
     <script>
-        $(function () {
+{#        $(function () {#}
```
what's this?

nickpell replied:
commented out to expose ```reportModule``` for console debugged.  removed the commented out code.

czue commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -1,24 +1,76 @@
+
+
+        var select2s = $('.choice_filter');
+        for(var i = 0; i < select2s.length; i++) {
```
how much of this was copy/pasted from [dynamic_choice_list.js](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports_core/templates/reports_core/filters/dynamic_choice_list_filter/dynamic_choice_list.js#L2) (and is there any way to share more code?)

czue commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,96 @@ var ReportModule = (function () {
+                        StaticChoiceFilter: 'select_value',
+                        StaticDatespanFilter: 'date_range'
+                    };
+                    for(var docType in docTypeToField) {
```
this is the same iterating over an object problem from the other PR (also semi-funny indentation)

nickpell replied:
fixed (not following about the indentation however)

czue commented:
still didn't review front end super closely, still looking good. i'm tempted to ask for tests but realize that would be a lot of plumbing

Commit:
cleanup commented out code

Commit:
cleanup for loops

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+                        if(filter.selectedValue.doc_type() == docType) {
+                            selectedFilterValues[filter.slug][field] = filter.selectedValue[field]();
+                        }
+                    });
```
Don't make functions within a loop.

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+                    var filter = filtersInReport[i];
+                    _.each(filter.selectedValue, function(observable) {
+                        observable.subscribe(changeSaveButton);
+                    });
```
Don't make functions within a loop.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+    function FilterConfig(report_id, reportId, filterValues, reportFilters, changeSaveButton) {
+        var self = this;
+
+        this.reportFilters = JSON.parse(JSON.stringify(reportFilters)) || {};
```
If reportFilters ever going to be undefined? If so, will stringify fail? If it'll always at least be an empty object, won't stringify make it `"{}"` and then parse make it `{}` so you don't need the `|| {}`?

nickpell replied:
got the wrong ordering, thanks

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+        var self = this;
+
+        this.reportFilters = JSON.parse(JSON.stringify(reportFilters)) || {};
+        _.each(this.reportFilters, function(filtersInReport, _id) {
```
Picky, but why `_id` not `id`

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
     }
 
-    function ReportConfig(report_id, display, availableReportIds, reportCharts, graph_configs, language) {
+    function FilterConfig(report_id, reportId, filterValues, reportFilters, changeSaveButton) {
```
`report_id` vs `reportId` is not a great way to distinguish between two variables

orangejenny replied:
This function is long and deals with complex data...how about a comment explaining what the parameters are, so future devs can get a quick overview of what it's doing? This style can work well: https://github.com/dimagi/Vellum/blob/master/src/core.js#L1265-L1278

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+        _.each(this.reportFilters, function(filtersInReport, _id) {
+            for (var i = 0; i < filtersInReport.length; i++) {
+                var filter = filtersInReport[i];
+                if (_id == report_id && filterValues.hasOwnProperty(filter.slug)) {
```
as best practice, how about `===` instead of `==`?

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+                for(var filterFieldsIndex = 0; filterFieldsIndex < filterFields.length; filterFieldsIndex++) {
+                    filter.selectedValue[filterFields[filterFieldsIndex]] = ko.observable(filter.selectedValue[filterFields[filterFieldsIndex]] || '');
+                }
+                filter.selectedValue.value = ko.observable(filter.selectedValue.value ? filter.selectedValue.value.join("\u001F") : '');
```
I'm a fan of commenting any use of hex codes

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+
+        // TODO - add user-friendly text
+        this.filterDocTypes = [null, 'AutoFilter', 'StaticDatespanFilter', 'CustomDataAutoFilter', 'StaticChoiceListFilter', 'StaticChoiceFilter'];
+        this.autoFilterTypes = ['case_sharing_group', 'location_id', 'username', 'user_id'];
```
Why not `_.keys(_filter_type_to_func)`? Is ordering important? Similarly, could you expose `doc_type_to_filter_class` and use it here rather than repeating the list?

nickpell replied:
Didn't spend a lot of time here since I'm hoping to change the UX

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+        this.date_range_options = ['last7', 'last30', 'lastmonth'];
+    }
+
+    function ReportConfig(report_id, display, availableReportIds,
```
so, I get that snake_case is data coming from python and camelCase is data created in JavaScript...but for someone who, say, just has to fix a bug in the javascript code, the mix of conventions is going to be a headache

nickpell replied:
I fixed some of the more egregious case inconsistencies introduced here.  I'll do a more through cleanup of this file in a separate PR.

orangejenny replied:
thank you!

orangejenny commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -1,24 +1,73 @@
+
+        var select2s = $('.choice_filter');
+        for(var i = 0; i < select2s.length; i++) {
+            var element = select2s.eq(i);
```
Stylistic, but the three lines above could be collapsed into `$('.choice_filter').each(function(element) {`

nickpell replied:
for some reason I get ```TypeError: element.val is not a function``` when I try that approach.  Played around with a few different iterations and this was the only one that worked for me (don't know why)

orangejenny replied:
huh, that's odd. don't have ideas offhand for why that doesn't work.

orangejenny replied:
my syntax was bad - element isn't the jQuery object, so you'd also need `var $element = $(element);`

(not expecting you to change anything...it was just bothering me to not know why that didn't work)

orangejenny commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -1,24 +1,73 @@
+            var element = select2s.eq(i);
+            var pageSize = 20;
+            // TODO: Ideally the separator would be defined in one place. Right now it is
+            //       also defined corehq.apps.userreports.reports.filters.CHOICE_DELIMITER
```
It's also in report-module.js twice...you could probably at least get it to a point where it's only defined once client-side and once server-side.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -1,24 +1,73 @@
+                // allowClear only respected if there is a non empty placeholder
+                placeholder: " ",
+                ajax: {
+                    // TODO - this is pretty hackish
```
this is...interesting :)

nickpell replied:
"we do interesting things here"

Commit:
only stringify dictionary

orangejenny commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -57,6 +106,60 @@
+                                            </select>
+                                        </th>
+                                        <th>
+                                            <div data-bind="visible: filter.selectedValue.doc_type() == 'AutoFilter'">
```
`===`

orangejenny replied:
same comment for the similar blocks below

Commit:
_id to id

orangejenny commented on corehq/apps/app_manager/templates/app_manager/module_view_report.html:
```diff
@@ -57,6 +106,60 @@
+                                            <div data-bind="visible: filter.selectedValue.doc_type() == 'StaticChoiceListFilter'">
+                                                <input type="text"
+                                                       class="choice_filter"
+                                                       style="width:300px"
```
Why does this need a custom style?

nickpell replied:
without it unpopulated input fields won't default to a friendly size:

<img width="322" alt="screen shot 2015-08-05 at 11 11 40 am" src="https://cloud.githubusercontent.com/assets/1757035/9089370/d6677856-3b62-11e5-84eb-4f3848fa82fa.png">

orangejenny replied:
Why is that a problem on this page and not for other select2s? (Wondering if this style should be somewhere more general.)

Commit:
add comment about hex code

Commit:
use ===

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +89,97 @@ var ReportModule = (function () {
+                        if(filter.selectedValue.doc_type() === docType) {
+                            selectedFilterValues[filter.slug][field] = filter.selectedValue[field]();
+                        }
+                    });
```
Don't make functions within a loop.

nickpell replied:
doing it anyway

Commit:
don't make function in loop

Commit:
define select2 separator in one place

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +92,96 @@ var ReportModule = (function () {
+
+                filter.dynamicFilterName = ko.computed(function () {
+                    return reportId() + '/' + filter.slug;
+                });
```
Don't make functions within a loop.

Commit:
add function comment

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +92,105 @@ var ReportModule = (function () {
+
+                filter.dynamicFilterName = ko.computed(function () {
+                    return reportId() + '/' + filter.slug;
+                });
```
Don't make functions within a loop.

Commit:
change reportId to selectedReportId

Commit:
change report_id to savedReportId

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +92,105 @@ var ReportModule = (function () {
+
+                filter.dynamicFilterName = ko.computed(function () {
+                    return selectedReportId() + '/' + filter.slug;
+                });
```
Don't make functions within a loop.

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/report-module.js:
```diff
@@ -89,7 +92,105 @@ var ReportModule = (function () {
+                        if(filter.selectedValue.doc_type() === docType) {
+                            selectedFilterValues[filter.slug][field] = filter.selectedValue[field]();
+                        }
+                    });
```
Don't make functions within a loop.

nickpell commented:
@orangejenny I fixed or commented on all the stylistic issues (and deleted some dimagimon ones) - let me know if you have any further comments.  There's no test coverage for changes since the last passing test.

orangejenny commented:
Were you able to do anything about this?

> how much of this was copy/pasted from dynamic_choice_list.js (and is there any way to share more code?)

Looks good to me once tests pass.

nickpell commented:
I don't think there's an easy way to share code.  The urls to autocomplete options are different and the elements are accessed differently (id lookup vs class lookup).

czue commented:
:tada:
